### PR TITLE
add function cache for completion

### PR DESCRIPTION
**Pull request by zhihuij**
hi guys, 

i have add a file cache for function match, all matched function list will write to a cache file (autoload/erlang_index), and when close and  open vim again, this cache file will be loaded,
and this will reduce the time for matching a module for its function list.

do you think it is valuable to accept this patch?

### CODE REVIEW

**Comment by zhihuij** [comment]
there is also a branch for Windows user, if anyone need it, check here: 
https://github.com/onlychoice/vimerl/tree/win-vimerl
you may need install some tools used by vimerl under Windows(grep , find...)

(just didn't find any proper place to put it, so put it here)

**Comment by jimenezrick** [comment]
I like the idea but we should think a way to handle the case when the content of the cache is old. With the current behavior, all is in memory so just restarting Vim fixes this. But using this cache file, when a module suffers modifications, after restarting Vim this cache continues been outdated.

So, putting this simple, _how to remove certain module from the cache (and file cache) to force a reload?_ Adding this command could be a nice solution.

I mean, I'm talking about the custom Erlang code of the user, which is changing, not the OTP code that rarely is modified.

I also found this bug when using the auto completion on a **user** module not compiled nor present in the cache file.

```
Error detected while processing function erlang_complete#Complete..<SNR>48_ErlangFindExternalFunc:                                                                 
line   42:
E716: Key not present in Dictionary: x
```

**Comment by jimenezrick** [comment]
Well, the simplest solution could be to add a command to purge **completely** the cache and the cache file. It could be a little bit too much sometimes but the user should use it just from time to time.

Simple to implement and effective, what do you think?

File: autoload/erlang_complete.vim [hunk 1]
```diff
@@ -131,6 +139,16 @@ function s:ErlangFindExternalFunc(module, base)
 		endif
 	endfor
 
+	let func_lists = s:modules_cache[a:module]
+	if len(func_lists) > 0
+		let tmp_cache = {a:module : func_lists}
+		" write cache to file
+		exe "redir! >> " . s:erlang_complete_cache_file
```
Comment (jimenezrick) [inline autoload/erlang_complete.vim:146]: Is it necessary the `!`? I think it suffices with `redir >>`.

File: autoload/erlang_complete.vim [hunk 2]
```diff
@@ -156,3 +174,18 @@ function s:ErlangFindLocalFunc(base)
...
+			let cache = eval(line)
+			for key in keys(cache)
+				" add module function list to the cache
+				let function_list = get(cache, key)
+				let s:modules_cache[key] = function_list
+			endfor
+		endfor
+	endif
+
+	return []
```
Comment (jimenezrick) [inline autoload/erlang_complete.vim:190]: I think the return is superfluous, isn't it?

**Comment by zhihuij** [comment]
i think remove certain module entry from the cache is better. we can provide a command with module name as a parameter, and this command will remove the entry from the cache and the cache file.

we can do this by first remove the entry from the cache in memory, and clear the cache file, and rewrite the cache to the cache file, how about this?

File: autoload/erlang_complete.vim (hunk 2 above)
Comment (zhihuij) [inline autoload/erlang_complete.vim:190]: yeah

**Comment by jimenezrick** [comment]
OK, simple and effective, I like you solution :smile:.

I am refactoring a little bit you code and fixing that bug I found out.

File: autoload/erlang_complete.vim (hunk 1 above)
Comment (zhihuij) [inline autoload/erlang_complete.vim:146]: i have checked the doc, and it says 
When [!] is omitted, and {file} exists, this command fails.
but when i remove the '!', it's works correctly. 

confused, any clues?

Comment (jimenezrick) [inline autoload/erlang_complete.vim:146]: I think the `!` only applies when using `>` and the file exists.

When using `>>` for appending in a file, it isn't necessary.

**Comment by zhihuij** [comment]
thank you very much.

you will add the command or should i submit another patch?

File: autoload/erlang_complete.vim (hunk 1 above)
Comment (zhihuij) [inline autoload/erlang_complete.vim:146]: got it, thanks

**Comment by jimenezrick** [comment]
If you want, submit another patch with the command.

**Comment by jimenezrick** [comment]
Let me finish the refactor and upload the code to the master branch. And add the patch there.

**Comment by zhihuij** [comment]
ok

**Comment by jimenezrick** [comment]
Done!, everything merged. If you want, add the `:ErlangPurgeModule` command or something like that.

**Comment by jimenezrick** [comment]
I am already writing the  `:ErlangPurgeCache` command.

**Pull request merged**